Carbon/Functions/Get-Principal.ps1

```diff
@@ -0,0 +1,87 @@
+
+function Get-CPrincipal
+{
+    <#
+    .SYNOPSIS
+    INTERNAL.
+
+    .DESCRIPTION
+    INTERNAL.
+
+    .EXAMPLE
+    INTERNAL.
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory)]
+        # The principal to search for.
+        [DirectoryServices.AccountManagement.Principal]$Principal,
+
+        [Parameter(Mandatory)]
+        [scriptblock]$Filter
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+    
+    $principalTypeName = 'principal'
+    if( $Principal -is [DirectoryServices.AccountManagement.UserPrincipal] )
+    {
+        $principalTypeName = 'user'
+    }
+    elseif( $Principal -is [DirectoryServices.AccountManagement.GroupPrincipal] )
+    {
+        $principalTypeName = 'group'
+    }
+    
+    Write-Timing 'Get-CPrincipal'
+    Write-Timing ('                {0}' -f $principalTypeName)
+
+    $searcher = New-Object 'DirectoryServices.AccountManagement.PrincipalSearcher' $Principal
+    try
+    {
+        $principals = @()
+
+        $maxTries = 100
+        $tryNum = 0
+        while( $tryNum++ -lt $maxTries )
+        {
+            $numErrorsBefore = $Global:Error.Count
+            try
+            {
+                Write-Timing ('                [{0,3} of {1}]  FindAll()  Begin' -f $tryNum,$maxTries)
+                $principals = 
+                    $searcher.FindAll() |
+                    Where-Object -FilterScript $Filter
+                Write-Timing ('                                FindAll()  End')
+                break
+            }
+            catch
+            {
+                Write-Timing ('                                FindAll()  Failed')
+                $_ | Out-String | Write-Debug 
+
+                $lastTry = $tryNum -ge $maxTries

```

Comment:
Will this evaluate to true? It looks like the `while` loop's last iteration will be `99` so this won't evaluate to true, e.g. `99 -ge 100`. I think you need to change the operator in the `while` condition to `-le`?

Reply:
It starts at `0`. After the `while` condition first evaluates, it's `1`. The `++` postfix operator increments the value _after_ comparing it to `$maxTries`. Inside the loop, on the last iteration, the value will be `100` not `99`.

Reply:
Perfect, wasn't sure if the compare happened before or after the increment.